### PR TITLE
[WIP] Add optimistic sharding mode

### PR DESCRIPTION
**Pull request by kennytm**
> WIP! Borrowing the CI for testing.

### What problem does this PR solve? 

Add a simple optimistic DDL lock mode. Enable with `dmctl set-ddl-lock-mode optimistic`.

### Check List 

Tests 

 - Unit test

Code changes

Side effects

Related changes

 - Need to update the documentation
 - Need to be included in the release note

### CODE REVIEW

**Comment by kennytm** [comment]
ok i'm out of idea how the optimistic lock should fit into the existing system.

current pessimistic-only system is roughly like:

```go
// master 1
go {
    for {
        task := <-infoCh 
        lock := FindLock(task)
        synced := lock.TrySync(task)
        if synced {
            opCh <- NewOperation(lock)
        }
    }
}    

// master 2
go {
    for {
        op := <-opCh 
        lock := FindLock(op)
        for _, source := range lock.Sources {
            if source == lock.Owner {
                clientCh <- "execute " + lock.DDLs
            } else {
                clientCh <- "skip " + lock.DDLs 
            }
        }
        RemoveLock(op)
    }
}

// client 
go {
    ...
    infoCh <- DDLs 
    action := <-clientCh  // (reader responsible for deleting the etcd keys??)
    ...
}
```

supporting optimistic-lock should be simply changing:

```go
// master 1
go {
    for {
        task := <-infoCh 
        lock := FindLock(task)
        newDDLs, err := lock.TrySync(task)       // <-- changed 
        if err != nil {                          
            opCh <- NewOperation(task, newDDLs)
        }
    }
}    

// master 2
go {
    for {
        op := <-opCh 
        lock := FindLock(op)
        for _, source := range lock.Sources {
            if source == lock.Owner {
                clientCh <- "execute " + op.DDLs // <-- changed
            } else {
                clientCh <- "skip"               // <-- changed 
            }
        }
        if lock.Synced() {                       // <-- changed 
            RemoveLock(op)
        }
    }
}

// client 
go {
    ...
    infoCh <- DDLs 
    action := <-clientCh
    ...
}
```

but i can't get the client to accept the modified DDLs 😕

**Comment by csuzhangxc** [comment]
in #522 I change only the leader of the master to coordinate shard DDL in pessimistic mode.

In pessimistic mode, the Info should be deleted by the worker, the Operation should be deleted by the  master, you can see _pkg/pessimism/doc.go_.

File: pkg/shardddl/pessimism/keeper.go [hunk 1]
```diff
@@ -18,24 +18,56 @@ import (
...
+	// upstream schemas.
+	LockModeOptimistic
 )
 
 // LockKeeper used to keep and handle DDL lock conveniently.
 // The lock information do not need to be persistent, and can be re-constructed from the shard DDL info.
 type LockKeeper struct {
 	mu    sync.RWMutex
 	locks map[string]*Lock // lockID -> Lock
+	mode  LockMode
```
Comment (csuzhangxc) [inline pkg/shardddl/pessimism/keeper.go]: Do we need to set a different `mode` for each migration task?

If so, maybe we can add a config item in task.yaml?

File: pkg/shardddl/optimism/lock.go [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+		}
+	}
+
+	l.joined = newJoined
+
+	// FIXME: Compute DDLs through schema diff instead of propagating DDLs directly.
+	cmpRes, err := oldJoined.Compare(newJoined)
+	l.synced = err == nil && cmpRes == 0
+	if err != nil || cmpRes == 0 {
+		ddls = nil
```
Comment (csuzhangxc) [inline pkg/shardddl/optimism/lock.go]: set `ddls = nil` for `err == nil && cmpRes == 0` (`l.synced = true`)?

**Comment by csuzhangxc** [comment]
implemented in https://github.com/pingcap/dm/pull/568.

**Pull request closed without merging**